### DISCUSSION TITLE
Why is the 'runs-on' not supported when calling reusable workflow?

### DISCUSSION THREAD

**Discussion by amoravek** [community/community#25783] [Actions]
<p>According to the documentation, reusable workflow callers do not support the ‘runs-on’ instruction. How is it possible to reuse workflows on BOTH github-hosted AND self-hosted runners, possibly in a single workflow? The documentation says that the self-hosted runners are supported through the caller’s context…but how do I do that, while having the runs-on instruction hardcoded inside the reusable workflow?</p>
<p>I feels more convenient to have the runs-on instruction set by the CALLER and not the reusable workflow - it does not make any sense to me? Do I get it wrong?</p>
<p>Thank you for any clarification.</p>

**Comment by jsoref** [comment, marked as answer]
<p>Here’s an example (tested, working):</p>
<pre><code class="lang-auto">commit 08e91da4450050f3ebbd53f90944e400e490bd55
Author: Josh Soref &lt;jsoref@users.noreply.github.com&gt;
Date:   Sat Dec 18 23:48:04 2021 -0500

    Run something?

diff --git a/.github/workflows/runs-on.yml b/.github/workflows/runs-on.yml
new file mode 100644
index 0000000..f888866
--- /dev/null
+++ b/.github/workflows/runs-on.yml
@@ -0,0 +1,16 @@
+name: callee
+on:
+  workflow_call:
+    inputs:
+      platform:
+        required: true
+        description: platform
+        type: string
+
+jobs:
+  run:
+    runs-on: ${{ inputs.platform }}
+    steps:
+    - name: bash-version
+      shell: bash
+      run: bash --version
diff --git a/.github/workflows/runs-places.yml b/.github/workflows/runs-places.yml
new file mode 100644
index 0000000..b8c322e
--- /dev/null
+++ b/.github/workflows/runs-places.yml
@@ -0,0 +1,10 @@
+name: caller
+on:
+  push:
+
+jobs:
+  runner:
+    name: run
+    uses: jsoref/reusable-runs-on/.github/workflows/runs-on.yml@main
+    with:
+      platform: ubuntu-18.04
</code></pre>
<p>output:</p>
<pre><code class="lang-auto">Current runner version: '2.285.1'
Operating System
  Ubuntu
  18.04.6
  LTS
Virtual Environment
Virtual Environment Provisioner
GITHUB_TOKEN Permissions
Secret source: Actions
Prepare workflow directory
Prepare all required actions
1s
Run bash --version
GNU bash, version 4.4.20(1)-release (x86_64-pc-linux-gnu)
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later &lt;http://gnu.org/licenses/gpl.html&gt;

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
</code></pre>
<pre><code class="lang-auto">commit d1ea2196f1f4a3d43843e424a8dd8763b44da405
Author: Josh Soref &lt;jsoref@users.noreply.github.com&gt;
Date:   Sat Dec 18 23:49:25 2021 -0500

    20!

diff --git a/.github/workflows/runs-places.yml b/.github/workflows/runs-places.yml
index b8c322e..8c92396 100644
--- a/.github/workflows/runs-places.yml
+++ b/.github/workflows/runs-places.yml
@@ -7,4 +7,4 @@ jobs:
     name: run
     uses: jsoref/reusable-runs-on/.github/workflows/runs-on.yml@main
     with:
-      platform: ubuntu-18.04
+      platform: ubuntu-20.04
</code></pre>
<p>Output:</p>
<pre><code class="lang-auto">Current runner version: '2.285.1'
Operating System
  Ubuntu
  20.04.3
  LTS
Virtual Environment
Virtual Environment Provisioner
GITHUB_TOKEN Permissions
Secret source: Actions
Prepare workflow directory
Prepare all required actions
1s
Run bash --version
GNU bash, version 5.0.17(1)-release (x86_64-pc-linux-gnu)
Copyright (C) 2019 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later &lt;http://gnu.org/licenses/gpl.html&gt;

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
</code></pre>

**Comment by amoravek** [comment]
<p>thank you, jsoref, it works. however it assumes that you have write access to the reusable workflow’s repository which is not always the case. the true reusability comes with the possibility of running reusable workflows wherever (github- and/or self-hosted) i need without any modification. that doesn’t seem to be possible now. or am i wrong?</p>

**Comment by jsoref** [comment]
<p>You aren’t wrong. In the interim, make a pull request to the repository offering the reusable workflow and point your consumer to your fork.</p>
<p>I’m not saying this is ideal, but, I actually can’t think of a better way to implement this. Imagine a reusable workflow with 13 jobs, some subset of which should probably be run on one class of system and another subset on a different subset. Or imagine that the author is in the process of adding/removing jobs.</p>
<p>I suppose GitHub could allow specifying runs-on from the caller based on the name of each nested job, but then that fails if the callee changes their jobs…</p>

**Comment by amoravek** [comment]
<p>you’re right, there are many conflicting scenarios i haven’t thought of. nevertheless, i am happy with the solution you’ve provided. thank you.</p>
